Userland/Libraries/LibWeb/Streams/WritableStream.h

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2023, Matthew Olsson <mattco@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Forward.h>
+#include <AK/SinglyLinkedList.h>
+#include <LibJS/Forward.h>
+#include <LibWeb/Bindings/PlatformObject.h>
+#include <LibWeb/Forward.h>
+#include <LibWeb/WebIDL/Promise.h>
+
+namespace Web::Streams {
+
+// https://streams.spec.whatwg.org/#pending-abort-request
+struct PendingAbortRequest {
+    // https://streams.spec.whatwg.org/#pending-abort-request-promise
+    // A promise returned from WritableStreamAbort
+    JS::NonnullGCPtr<WebIDL::Promise> promise;
+
+    // https://streams.spec.whatwg.org/#pending-abort-request-reason
+    // A JavaScript value that was passed as the abort reason to WritableStreamAbort
+    JS::Value reason;
+
+    // https://streams.spec.whatwg.org/#pending-abort-request-was-already-erroring
+    // A boolean indicating whether or not the stream was in the "erroring" state when WritableStreamAbort was called, which impacts the outcome of the abort request
+    bool was_already_erroring;
+};
+
+// https://streams.spec.whatwg.org/#writablestream
+class WritableStream final : public Bindings::PlatformObject {
+    WEB_PLATFORM_OBJECT(WritableStream, Bindings::PlatformObject);
+
+public:
+    enum class State {
+        Writable,
+        Closed,
+        Erroring,
+        Errored,
+    };
+
+    static WebIDL::ExceptionOr<JS::NonnullGCPtr<WritableStream>> construct_impl(JS::Realm& realm, Optional<JS::Handle<JS::Object>> const& underlying_sink);
+
+    virtual ~WritableStream() = default;
+
+    bool locked();
+
+    bool backpressure() const { return m_backpressure; }
+    void set_backpressure(bool value) { m_backpressure = value; }
+
+    JS::GCPtr<WebIDL::Promise> close_request() const { return m_close_request; }
+    void set_close_request(JS::GCPtr<WebIDL::Promise> value) { m_close_request = value; }
+
+    JS::GCPtr<JS::Object> controller() { return m_controller; }
+    void set_controller(JS::GCPtr<JS::Object> value) { m_controller = value; }
+
+    JS::GCPtr<WebIDL::Promise> in_flight_write_request() const { return m_in_flight_write_request; }
+    void set_in_flight_write_request(JS::GCPtr<WebIDL::Promise> value) { m_in_flight_write_request = value; }
+
+    JS::GCPtr<WebIDL::Promise> in_flight_close_request() const { return m_in_flight_close_request; }
+    void set_in_flight_close_request(JS::GCPtr<WebIDL::Promise> value) { m_in_flight_close_request = value; }
+
+    Optional<PendingAbortRequest>& pending_abort_request() { return m_pending_abort_request; }
+    void set_pending_abort_request(Optional<PendingAbortRequest>&& value) { m_pending_abort_request = move(value); }
+
+    State stream_state() const { return m_state; }
+    void set_stream_state(State value) { m_state = value; }
```

Comment:
This should be called `state()` and `set_state()`. Getter names always mirror their member names by default.

Reply:
I named it this way because otherwise it shadows `JS::Cell::{state,set_state}`. Is that not a problem?

Reply:
That's fine, it's not a virtual method so we can (and already do - see `Userland/Libraries/LibWeb/Fetch/Infrastructure/FetchController.h` for example) shadow it in subclasses. Anything messing with cell state only operates on `Cell` pointers anyway and has no idea this subclass method of the same name exists.